### DISCUSSION TITLE
Can I customize keep-alive settings?

### DISCUSSION THREAD

**Discussion by dayvejones** [tokio-rs/axum#2939] [Q&A] [answered]
### Summary

Hello! 

I assume the axum(hyper) uses the following things:
https://docs.rs/hyper/latest/hyper/server/conn/http1/struct.Builder.html#method.keep_alive
https://docs.rs/socket2/latest/socket2/struct.Socket.html#method.set_tcp_keepalive

Are there any elegant ways to control the option **keep_alive**(enable/disable) setting globally when the app starts? Can I pass config for Hyper when use **axum::serve(listener, app)** ?

Are there any ways to control the option **keep_alive** setting in middleware only for some connections?

For example, actix:

https://actix.rs/docs/server/#keep-alive

```
use actix_web::{http::KeepAlive, HttpServer};
use std::time::Duration;

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    // Set keep-alive to 75 seconds
    let _one = HttpServer::new(app).keep_alive(Duration::from_secs(75));

    // Use OS's keep-alive (usually quite long)
    let _two = HttpServer::new(app).keep_alive(KeepAlive::Os);

    // Disable keep-alive
    let _three = HttpServer::new(app).keep_alive(None);

    Ok(())
}
```

Thanks!

### axum version

0.7.6

**Comment by jplatte** [comment, marked as answer]
There is no way with `axum::serve` right now (feel free to open a feature request), but if you follow the [`serve-with-hyper` example](https://github.com/tokio-rs/axum/blob/v0.7.x/examples/serve-with-hyper/src/main.rs), you can move the `server::conn::auto::Builder` to a new binding, and call [`.http1()`](https://docs.rs/hyper-util/latest/hyper_util/server/conn/auto/struct.Builder.html#method.http1)[`.keep_alive()`](https://docs.rs/hyper/latest/hyper/server/conn/http1/struct.Builder.html#method.keep_alive) on it before `serve_connection_with_upgrades`.

**Reply by dayvejones** [reply]
Thanks!

**Reply by IvanUkhov** [reply]
This is not something one can combine with `Serve::with_graceful_shutdown`, right?

**Reply by mladedav** [reply]
No, but there's [graceful shutdown in hyper](https://hyper.rs/guides/0.14/server/graceful-shutdown/).

**Reply by baxterjo** [reply]
It doesn't look like this solution provides the ability to change the keep-alive timeout duration as the question asks. Is there a way to set this through some other method?

**Comment by IvanUkhov** [comment]
To understand the situation better, is it then the case that the server is losing one file descriptor each time a connection is not closed properly? So over time, the server is expected to run out of file descriptors. And it applies to all servers using Axum (without any auxiliary tweaks, such as the one above). Is this correct?

**Reply by jplatte** [reply]
This has nothing to do with the original question. Please post a separate discussion.

**Comment by gmarti** [comment]
What about [#2938 ](https://github.com/tokio-rs/axum/pull/3852)
